File: backend/app/schemas/export.py
from __future__ import annotations

import uuid
from typing import Annotated, Literal

from pydantic import BaseModel, Field


ExportSectorCount = Literal[12, 16, 36]


class CSVExportRequest(BaseModel):
    column_ids: list[uuid.UUID] = Field(default_factory=list)
    exclude_flags: list[uuid.UUID] = Field(default_factory=list)
    resample: str | None = None


class WAsPTabExportRequest(BaseModel):
    speed_column_id: uuid.UUID
    direction_column_id: uuid.UUID
    exclude_flags: list[uuid.UUID] = Field(default_factory=list)
    num_sectors: ExportSectorCount = 12
    speed_bin_width: Annotated[float, Field(default=1.0, gt=0, le=10)] = 1.0


class IEAJSONExportRequest(BaseModel):
    column_ids: list[uuid.UUID] = Field(default_factory=list)
    exclude_flags: list[uuid.UUID] = Field(default_factory=list)
    resample: str | None = None


class OpenwindExportRequest(BaseModel):
    column_ids: list[uuid.UUID] = Field(default_factory=list)
    exclude_flags: list[uuid.UUID] = Field(default_factory=list)
    resample: str | None = None

File: frontend/src/types/export.ts
export type ExportFormat = "csv" | "wasp-tab" | "iea-json" | "openwind";

export interface CSVExportRequest {
  column_ids?: string[];
  exclude_flags?: string[];
  resample?: string;
}

export interface WAsPTabExportRequest {
  speed_column_id: string;
  direction_column_id: string;
  exclude_flags?: string[];
  num_sectors?: 12 | 16 | 36;
  speed_bin_width?: number;
}

export interface IEAJSONExportRequest {
  column_ids?: string[];
  exclude_flags?: string[];
  resample?: string;
}

export interface OpenwindExportRequest {
  column_ids?: string[];
  exclude_flags?: string[];
  resample?: string;
}

export interface ExportDownload {
  blob: Blob;
  fileName: string;
  contentType: string;
}